### DISCUSSION TITLE
PKCE vs Code confusion

### DISCUSSION THREAD

**Discussion by hyperknot** [zitadel/zitadel#4999] [Q&A] [answered]
The tutorials say:
- We recommend that you use [Authorization Code](https://zitadel.com/docs/apis/openidoauth/grant-types#authorization-code) in combination with [Proof Key for Code Exchange](https://zitadel.com/docs/apis/openidoauth/grant-types#proof-key-for-code-exchange) for all web applications.
- As the requests from your application to ZITADEL are made on NextJS serverside, you can select CODE in the next step.

What I don't understand is that on the screen I can only select either PKCE or Code. How can I select a combination? 
Also, there is **no next step**.

<img width="1255" alt="image" src="https://user-images.githubusercontent.com/494223/211152695-817de156-0185-4f35-8d6a-4e922548838a.png">

Can you clarify these points? Was the UI changed since the docs were written?

**Comment by livio-a** [comment, marked as answer]
Hey @hyperknot 

Maybe we need to clarify the docs a bit as PKCE is an addition to the regular Code Flow. So if you select PKCE on that screen, you configuring your app to do a Code Flow, but with code_challenge / code_verifier instead of a client_secret (like PKCE for native and user agent apps).
So the...
> As the requests from your application to ZITADEL are made on NextJS serverside, you can select CODE in the next step.

... is written as the nodejs expects to also send a client_secret even with PKCE (like in OAuth2.1 required for web apps), which ZITADEL is currently no capable. 
@peintnermax can you confirm this?
We'll require client authorization for web apps (with client_id / client_secret) with PKCE as well this in the future: https://github.com/zitadel/oidc/issues/254

Using regular code flow for web apps without PKCE is totally safe.

**Comment by hyperknot** [comment]
So if I understand correctly, on a frontend-only app (for example create-react-app), I have to select PKCE on that screen, because there are no secrets.

If I have a backend on the other hand (like with Next.js), I can use CODE, because it can handle secrets, right?

The middle block you wrote I didn't understand, are you discussing that in the future you'd send both PKCE and client_secret from backend apps?

**Reply by livio-a** [reply]
Yes you're correct with everything.

OAuth 2.1 (currently still in draft) will require every app to use PKCE. And backend apps will need to send both.
Currently some already do this, but ZITADEL only handles PKCE without client_secret. We'll change this in the future.

**Reply by hyperknot** [reply]
Thanks a lot for the explanation.
